### DISCUSSION TITLE
Can't get Elixir LS to respect root_spec

### DISCUSSION THREAD

**Discussion by dinvlad** [LazyVim/LazyVim#1971] [Q&A] [answered]
I apologize if this is not the right avenue to ask - I'm facing an issue where the Elixir language server improperly replaces spaces with parentheses in all function invocations in my codebase on save. It's happening because `elixir-ls` detects the root directory to be the Git repository, while my `.formatter.exs` file is in its child directory for the associated Elixir app (this is the default configuration that Elixir Phoenix framework creates). As a result, Elixir LS ignores the formatter options.

I tried to get it to detect the root dir to be the app directory where I start Neovim, by setting
`vim.g.root_spec = { "cwd" }`. I then verified that this option is indeed picked up by Neovim, with `:LazyRoot`.
However, `:LspInfo` still returns the parent directory:
```
Language client log: /Users/<user>/.local/state/nvim/lsp.log
 Detected filetype:   elixir
 
 1 client(s) attached to this buffer: 
 
 Client: elixirls (id: 1, bufnr: [63])
 	filetypes:       elixir, eelixir, heex, surface
 	autostart:       true
 	root directory:  /Users/<user>/code/phoenix-apps
 	cmd:             elixir-ls
```
Unfortunately, I can't just put `.formatter.exs` into the parent directory, since it includes options that are specific to the child directory, and there could be multiple child directories corresponding to multiple apps in this monorepo.

`lsp.log` shows the following:
```
LSP logging initiated
.../vim/lsp/rpc.lua:734    "rpc"   "elixir-ls"     "stderr"        "Running /Users/<user>/.local/share/nvim/mason/packages/elixir-ls/launch.sh\nPreferred shell is zsh, relaunching\nLooking for ASDF install\nASDF not found\nLooking for rtx executable\nrtx not found\n"
.../vim/lsp/rpc.lua:734    "rpc"   "elixir-ls"     "stderr"        "Installing ElixirLS release v0.17.7\n"
.../vim/lsp/rpc.lua:734    "rpc"   "elixir-ls"     "stderr"        "Running in /Users/<user>/code/phoenix-apps\n"
.../vim/lsp/rpc.lua:734    "rpc"   "elixir-ls"     "stderr"        "Install complete\n"
...lsp/handlers.lua:537     'No mixfile found in project. To use a subdirectory, set `elixirLS.projectDir` in your settings. Looked for mixfile at "/Users/<user>/code/phoenix-apps/mix.exs"'
```
which seemingly suggests I should set `elixirLS.projectDir`, but I think that option applies only to VS Code (in which this very same formatting config works fine, and the LS is in fact started inside the child app directory, when I open it in VS Code).

Thanks for any help!

P.S. The only relevant answer I could find (and where I took the suggestion about root_spec) is https://github.com/LazyVim/LazyVim/issues/1912#issuecomment-1786880134

**Comment by dpetka2001** [comment, marked as answer]
This happens because of the `root_dir` detection by `elixirls`. You can also see it in the lspconfig [docs](https://github.com/neovim/nvim-lspconfig/blob/37457f268af5cd6765e589b0dcd7cbd192d8da00/lua/lspconfig/server_configurations/elixirls.lua#L6-L8). It first detects if a `.git` directory exists and then the rest. That's why in your case the `root_dir` of your project is the same as where the `.git` directory resides. You can change this kind of behavior if i'm not mistaken. Take a look at #1307 for a similar problem.

**Comment by dinvlad** [comment]
Thanks for the help! I ended up setting up an umbrella project using `mix new --umbrella` and moving my app under `apps/`, which then fixed all issues, and then I didn't have to adjust `root_dir`. Now ElixirLS + LazyVim correctly figure out the right formatter to use depending on the file I'm editing.
